### PR TITLE
Feature hot reload

### PR DESCRIPTION
**Pull request by Pandabear1125**
Implements hot reloading. This enables Teensy to be reconfigured after startup (say if Hive goes down and comes back up and wants to configure Teensy differently). 

This also has the nice side effect of making Hive resilient to Teensy crashes. Teensy can now die and come back to life and Hive still can maintain a connection with it. 

Teensy is also now resilient to Hive crashes, since Hive can go down and come back up, triggering another configure.

### CODE REVIEW

File: src/comms/config_layer.cpp [hunk 1]
```diff
@@ -1,5 +1,17 @@
...
+/// @brief This resets the whole processor and kicks it back to program entry (teensy4/startup.c)
+/// @param void void
+/// @note Dont abuse this function, it is not to be used lightly
+extern "C" void reset_teensy(void) {
+    // Register information found in the NXP IM.XRT 1060 reference manual
+	SRC_GPR5 = 0x0BAD00F1;
+    // Register information found in the Arm-v7-m reference manual
+	SCB_AIRCR = 0x05FA0004;
+    // loop go brr
+	while (1);
```
Comment (ashleycody345) [inline src/comms/config_layer.cpp:12]: I don't think there's any way for us to actually enter this loop, but do we know what happens if we do for any reason? Firmware should just lock up entirely, if I'm not mistaken.

This loop is absolutely how this should be implemented, I'm just curious to know if we have any failsafes if this loop ever does get entered for some reason.

File: src/main.cpp [hunk 2]
```diff
@@ -156,6 +155,16 @@ int main() {
         CommsPacket* incoming = comms.get_incoming_packet();
         CommsPacket* outgoing = comms.get_outgoing_packet();
 
+        // check whether this packet is a config packet
+        if (incoming->raw[3] == 1) {
```
Comment (ashleycody345) [inline src/main.cpp:159]: This looks like it relies on the current implementation for how HID handles sending/receiving config packets. How much would this need to be edited to include support for Ethernet down the road- would we really only need to change this one line and make sure that we use the appropriate configure procedure?

**Review by ashleycody345 — APPROVED** [review]
I had a few minor concerns about how certain functions might work, so I left some comments asking for general information. However, I don't think there is anything wrong with how this is implemented at the moment. Great work!

File: src/comms/config_layer.cpp (hunk 1 above)
Comment (Pandabear1125) [reply src/comms/config_layer.cpp:12]: As discussed, the system reset call isnt instant. It may take multiple cycles to complete. The loop is to catch the program so it doesnt do anything else in that time it takes to reset.

File: src/main.cpp (hunk 2 above)
Comment (Pandabear1125) [reply src/main.cpp:159]: This line would only need to be updated to check whether the incoming generic packet is marked as config. However that is done is left to the implemention. Essentially this check is if (incoming is config)

File: src/comms/config_layer.cpp [hunk 3]
```diff
@@ -125,8 +125,8 @@ void ConfigLayer::reconfigure(HIDLayer* comms) {
 
     // update stored config, msg if successful
     Serial.println("Attempting to store config...");
-    if (store_config()) Serial.println("Config successfully stored in /config.pack");
-    else Serial.println("Config not successfully stored (is an SD card inserted?)");    // not a fatal error, can still return
+    if (store_config()) Serial.println("\tConfig successfully stored in /config.pack\n");
+    else Serial.println("\tConfig not successfully stored (is an SD card inserted?)\n");    // not a fatal error, can still return
```
Comment (ashleycody345) [inline src/comms/config_layer.cpp:129]: This is a `Serial.println` call and not a `Serial.printf` one - is the `\n` necessary?

**Review by ashleycody345 — COMMENTED** [review]
Looks fine overall, just small questions

File: src/comms/config_layer.cpp (hunk 3 above)
Comment (Pandabear1125) [reply src/comms/config_layer.cpp:129]: Yeah it makes the output easier to read

**Comment by Pandabear1125** [comment]
Tested on sentry and infantry, no explosions

File: src/comms/config_layer.cpp [hunk 4]
```diff
@@ -1,5 +1,17 @@
 #include "config_layer.hpp"
 
+/// @brief This resets the whole processor and kicks it back to program entry (teensy4/startup.c)
+/// @param void void
```
Comment (StudentKyle) [inline src/comms/config_layer.cpp:4]: Haven't seen this before, but worth documenting since that is just weird historical C that isn't present in C++.
```suggestion
/// @param void specify no arguments (needed in C)
```

File: src/comms/config_layer.hpp [hunk 5]
```diff
@@ -26,8 +26,8 @@
 // lastly, stores bytes of subsec_sizes array
 #define CONFIG_PATH "/config.pack"
 
-// define CONFIG_OFF_ROBOT macro when running off of real robot (testing firmware away from actual robot)
-// #define CONFIG_OFF_ROBOT 
+// define DISREGARD_REF_CONFIG_CHECK macro when running off of real robot (testing firmware away from actual robot)
+#define DISREGARD_REF_CONFIG_CHECK
```
Comment (StudentKyle) [inline src/comms/config_layer.hpp:30]: `IGNORE_REF_CONFIG_SAFETY` or something similar with the word `SAFETY` in it sounds scarier, making it clear that disabling it has safety implications. Or `DISABLE` as well to be clearer that this is also disabling ref elsewhere (if I understand correctly).

File: src/comms/config_layer.cpp [hunk 6]
```diff
@@ -42,17 +54,96 @@ const Config* const ConfigLayer::configure(HIDLayer* comms) {
...
+    // Hive marks the config process as done once teensy sends a packet with info_bit == 0
+    // Hive then sends a packet back with info_bit == 0 as well, so ping until we get a non-config back
+    while (comms->get_incoming_packet()->raw[3] == 1) {
+        comms->ping();
+    }
 
     return &config;
 }
 
+void ConfigLayer::reconfigure(HIDLayer* comms) {
```
Comment (StudentKyle) [inline src/comms/config_layer.cpp:81]: configure and reconfigure can't have this much duplicate code between them for maintainability reasons. I'd say either make everything just call the one `configure` and make them do the same thing, or have reconfigure call the configure function inside of it, so you can have your before and after behavior still there (which seems to be the only difference).

File: src/comms/config_layer.hpp [hunk 7]
```diff
@@ -7,7 +7,7 @@
 
 #include <map>
 #include <string>
-#define CONFIG_LAYER_DEBUG
+// #define CONFIG_LAYER_DEBUG
```
Comment (StudentKyle) [inline src/comms/config_layer.hpp:10]: was this intentional?

**Review by StudentKyle — CHANGES_REQUESTED** [review]
Some feedback. Excited for this feature!

File: src/comms/config_layer.hpp (hunk 7 above)
Comment (Pandabear1125) [reply src/comms/config_layer.hpp:10]: no lol

File: src/comms/config_layer.hpp
Comment (StudentKyle) [inline src/comms/config_layer.hpp:1]: comms parameter is not documented in config_SD_init

https://github.com/CU-Robotics/firmware/blob/9dad580e276fbd394dab5e5f22d8569043108d60/src/comms/config_layer.hpp#L155

File: src/comms/config_layer.hpp [hunk 8]
```diff
@@ -26,9 +25,8 @@
 // lastly, stores bytes of subsec_sizes array
 #define CONFIG_PATH "/config.pack"
 
-// define CONFIG_OFF_ROBOT macro when running off of real robot (testing firmware away from actual robot)
-// #define CONFIG_OFF_ROBOT 
-
+// define DISABLE_REF_CONFIG_SAFETY_CHECK macro when running off of real robot (testing firmware away from actual robot)
+#define DISABLE_REF_CONFIG_SAFETY_CHECK
```
Comment (StudentKyle) [inline src/comms/config_layer.hpp:29]: remember to un-define this before merge.

**Review by StudentKyle — CHANGES_REQUESTED** [review]
Looks great! Just noticed an undocumented parameter, and also remember to un-define the disable safety check.

File: src/comms/config_layer.hpp (hunk 8 above)
Comment (Pandabear1125) [reply src/comms/config_layer.hpp:29]: oop thanks haha

**Review by StudentKyle — APPROVED** [review]
Good

**Comment by StudentKyle** [comment]
@Pandabear1125 can add the edge case documentation in another PR, but he says the tests passed, so merging.
